pinot-core/src/test/java/com/linkedin/pinot/core/data/readers/BackfillDateTimeRecordReaderTest.java

```diff
@@ -0,0 +1,250 @@
+    inputSchema = createPinotSchemaWithTimeSpec(timeFieldSpec);
+    inputRecordReader = new TestRecordReader(inputData, inputSchema);
+    dateTimeFieldSpec = new DateTimeFieldSpec("timestampInEpoch", DataType.LONG, "1:HOURS:EPOCH", "1:HOURS", DateTimeType.PRIMARY);
+    wrapperSchema = createPinotSchemaWrapperWithDateTimeSpec(inputSchema, dateTimeFieldSpec);
+    entries.add(new Object[] {
+        inputRecordReader, timeFieldSpec, dateTimeFieldSpec, wrapperSchema
+    });
+
+    // timeSpec in millisSinceEpoch, generate dateTimeFieldSpec in 5 minutesSinceEpoch
+    timeFieldSpec = new TimeFieldSpec(new TimeGranularitySpec(DataType.LONG, TimeUnit.MILLISECONDS, "Date"));
+    inputData = createTestDataWithTimespec(timeFieldSpec);
+    inputSchema = createPinotSchemaWithTimeSpec(timeFieldSpec);
+    inputRecordReader = new TestRecordReader(inputData, inputSchema);
+    dateTimeFieldSpec = new DateTimeFieldSpec("timestampInEpoch", DataType.LONG, "5:MILLISECONDS:EPOCH", "1:HOURS", DateTimeType.PRIMARY);
+    wrapperSchema = createPinotSchemaWrapperWithDateTimeSpec(inputSchema, dateTimeFieldSpec);
+    entries.add(new Object[] {
+        inputRecordReader, timeFieldSpec, dateTimeFieldSpec, wrapperSchema
+    });
+
+    // timeSpec in hoursSinceEpoch, dateTimeFieldSpec in millisSinceEpoch, override dateTimeFieldSpec in millisSinceEpoch
+    timeFieldSpec = new TimeFieldSpec(new TimeGranularitySpec(DataType.LONG, TimeUnit.HOURS, "Date"));
+    dateTimeFieldSpec = new DateTimeFieldSpec("timestampInEpoch", DataType.LONG, "1:MILLISECONDS:EPOCH", "1:HOURS", DateTimeType.PRIMARY);
+    inputData = createTestDataWithTimespec(timeFieldSpec, dateTimeFieldSpec);
+    inputSchema = createPinotSchemaWithTimeSpec(timeFieldSpec, dateTimeFieldSpec);
+    inputRecordReader = new TestRecordReader(inputData, inputSchema);
+    entries.add(new Object[] {
+        inputRecordReader, timeFieldSpec, dateTimeFieldSpec, inputSchema
+    });
+
+    // timeSpec in hoursSinceEpoch, dateTimeFieldSpec in hoursSinceEpoch, override dateTimeFieldSpec in millisSinceEpoch
+    timeFieldSpec = new TimeFieldSpec(new TimeGranularitySpec(DataType.LONG, TimeUnit.HOURS, "Date"));
+    dateTimeFieldSpec = new DateTimeFieldSpec("timestampInEpoch", DataType.LONG, "1:MILLISECONDS:EPOCH", "1:HOURS", DateTimeType.PRIMARY);
+    inputData = createTestDataWithTimespec(timeFieldSpec, dateTimeFieldSpec);
+    inputSchema = createPinotSchemaWithTimeSpec(timeFieldSpec, dateTimeFieldSpec);
+    inputRecordReader = new TestRecordReader(inputData, inputSchema);
+    dateTimeFieldSpec = new DateTimeFieldSpec("timestampInEpoch", DataType.LONG, "1:MILLISECONDS:EPOCH", "1:HOURS", DateTimeType.PRIMARY);
+    wrapperSchema = createPinotSchemaWithTimeSpec(timeFieldSpec, dateTimeFieldSpec);
+    entries.add(new Object[] {
+        inputRecordReader, timeFieldSpec, dateTimeFieldSpec, wrapperSchema
+    });
+
+    return entries.toArray(new Object[entries.size()][]);
+  }
```

Comment:
Do you want to add a test with multiple datetimefields?

Reply:
Added, thanks for catching this